### PR TITLE
mraa mock

### PR DESCRIPTION
**Pull request by alext-mkrs**
Here comes the mocking capability foundation + GPIO functionality implementation (without ISR) discussed in #531, #514 and #44 and done the way I deem the best (at the backend and as a separate platform - see my reasoning in #531).

The second commit enables mraa to build under MSYS2 in Windows and the mock platform was tested to work fine there. If nothing else, it solves the problem of autocompletion in Pythin IDEs, raised in #375 :smiley: I've personally tested it to work just fine under PyCharm, see screenshot below.

Your comments and suggestions are more than welcome. There are several things to note:

1. The way we used `BUILDARCH` in CMake, seems to be incorrect - we define it as "option" (i.e. as a bool), but then assign strings to it - so I fixed this to be a string from the beginning. I haven't split it out as a separate commit.
1. I've used 96 as the platform ID as I recall there was a suggestion to use 97 for something else recently.
1. To avoid cutting out the whole `spi.c` file under Windows I had to extract definitions from `spidev.h` to `spi_kernel_headers.h` in the way it was previously done for IIO and for files under `include/linux`. MSYS2 does not have `spidev.h` available at all.
1. I have tried to compile it using bare MinGW, but we rely on POSIX things a bit too much (e.g. `syslog()` all over the place), so it didn't work and would require quite some effort to enable it. MSYS2 on the other hand only adds one DLL dependency, which should be ok, given the tasks mraa mock on Windows is aimed to solve. As a bonus, MSYS2 provides nice complete UNIX-like environment and has a convenient package manager.

And here's mraa in all its glory running under Windows :sunglasses:

pycharm_win_autocompletion

### CODE REVIEW

File: api/mraa/spi_kernel_headers.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+#define SPI_IOC_RD_BITS_PER_WORD        _IOR(SPI_IOC_MAGIC, 3, __u8)
+#define SPI_IOC_WR_BITS_PER_WORD        _IOW(SPI_IOC_MAGIC, 3, __u8)
+
+#define SPI_IOC_RD_MAX_SPEED_HZ         _IOR(SPI_IOC_MAGIC, 4, __u32)
+#define SPI_IOC_WR_MAX_SPEED_HZ         _IOW(SPI_IOC_MAGIC, 4, __u32)
+
+#define SPI_IOC_RD_MODE32               _IOR(SPI_IOC_MAGIC, 5, __u32)
+#define SPI_IOC_WR_MODE32               _IOW(SPI_IOC_MAGIC, 5, __u32)
+
+#endif
```
Comment (arfoll) [inline api/mraa/spi_kernel_headers.h]: Can we have this file in include/ and not in the public API?

**Comment by arfoll** [comment]
I love this thanks @alext-mkrs! I've merged the first patch, the second patch just watining if I can move that header into include/ to make things a bit cleaner.

File: api/mraa/spi_kernel_headers.h (hunk 1 above)
Comment (alext-mkrs) [inline api/mraa/spi_kernel_headers.h]: Sure thing, can do that easily - which location do you think would be better: `include/` of `include/linux`?

**Comment by alext-mkrs** [comment]
Glad to read that :smiley:. If you mean you can just move that during the merge, then by all means please do, I picked the location by using the iio include as the example, but I indeed spent some time thinking if `include` or `include/linux` is not a better one.

File: api/mraa/spi_kernel_headers.h (hunk 1 above)
Comment (arfoll) [inline api/mraa/spi_kernel_headers.h]: yeah include/linux is perfect for this

**Comment by arfoll** [comment]
ok I'll move it when I merge it no problems. yeah the iio is kept in there because we use the channel types and iio_event_data and I dind't want to uselessly redefine all that stuff. The example to follow is probably the i2c.h one.

**Comment by alext-mkrs** [comment]
Got it. Ok, I have this change on my branch, but haven't pushed - if you want me to do that (including rebasing to the latest master), please just tell.

**Comment by arfoll** [comment]
Well If you've done it then yes go ahead :)

**Comment by alext-mkrs** [comment]
Done

**Comment by arfoll** [comment]
Thanks again! Merged.

**Pull request closed without merging**